File: src/main/java/com/youngfeng/ideaplugin/java/siyeh_ig/psiutils/SerializationUtils.java
package com.youngfeng.ideaplugin.java.siyeh_ig.psiutils;

import com.intellij.psi.*;
import com.intellij.psi.util.InheritanceUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.kotlin.psi.KtClass;
import org.jetbrains.kotlin.psi.KtClassOrObject;
import org.jetbrains.kotlin.psi.KtSuperTypeListEntry;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

@SuppressWarnings({"UnusedDeclaration"})
public class SerializationUtils {

    private static final String SERIALIZABLE_CLASS_NAME   = "java.io.Serializable";
    private static final String EXTERNALIZABLE_CLASS_NAME = "java.io.Externalizable";

    private SerializationUtils() {
        // Nothing to do
    }

    public static boolean isSerializable(@Nullable PsiClass psiClass) {
        if (psiClass == null) {
            return false;
        }

		final PsiClass serializable = ClassUtils.findPsiClass(psiClass.getManager(), SERIALIZABLE_CLASS_NAME);

		return InheritanceUtil.isInheritorOrSelf(psiClass, serializable, true);
    }

	public static boolean isSerializable(@Nullable KtClass psiClass) {
		if (psiClass == null) {
			return false;
		}

		return isImplementingSerializable(psiClass);
	}

	public static boolean isImplementingSerializable(KtClassOrObject classOrObject) {
		List<KtSuperTypeListEntry> list = classOrObject.getSuperTypeListEntries();

		AtomicBoolean result = new AtomicBoolean(false);
		list.forEach(ktSuperTypeListEntry -> {
			if ("Serializable".equals(ktSuperTypeListEntry.getText()) || "java.io.Serializable".equals(ktSuperTypeListEntry.getText())) {
				result.set(true);
			}
		});

		return result.get();
	}

    public static boolean isExternalizable(@NotNull PsiClass psiClass) {
		final PsiClass serializable = ClassUtils.findPsiClass(psiClass.getManager(), EXTERNALIZABLE_CLASS_NAME);

        return InheritanceUtil.isInheritorOrSelf(psiClass, serializable, true);
    }

    public static boolean isDirectlySerializable(@NotNull PsiClass psiClass) {
        final PsiReferenceList implementsList = psiClass.getImplementsList();

        if (implementsList != null) {
	        for (final PsiJavaCodeReferenceElement aInterfaces : implementsList.getReferenceElements()) {
		        final PsiClass implemented = (PsiClass) aInterfaces.resolve();

		        if (implemented != null && SERIALIZABLE_CLASS_NAME.equals(implemented.getQualifiedName())) {
			        return true;
		        }
	        }

        }
        return false;
    }

    public static boolean hasReadObject(@NotNull PsiClass psiClass) {
	    for (final PsiMethod psiMethod : psiClass.getMethods()) {
		    if (isReadObject(psiMethod)) {
			    return true;
		    }
	    }

        return false;
    }

    public static boolean hasWriteObject(@NotNull PsiClass psiClass) {
	    for (final PsiMethod psiMethod : psiClass.getMethods()) {
		    if (isWriteObject(psiMethod)) {
			    return true;
		    }
	    }

        return false;
    }

    public static boolean isReadObject(@NotNull PsiMethod psiMethod) {
	    return isReadWriteObject(psiMethod, "readObject", "java.io.ObjectInputStream");
    }

	public static boolean isWriteObject(@NotNull PsiMethod psiMethod) {
		return isReadWriteObject(psiMethod, "writeObject", "java.io.ObjectOutputStream");
    }

    public static boolean isReadResolve(@NotNull PsiMethod method) {
	    return isReadWriteResolve(method, "readResolve");
    }

	public static boolean isWriteReplace(@NotNull PsiMethod method) {
		return isReadWriteResolve(method, "writeReplace");
    }

	private static boolean isReadWriteObject(@NotNull PsiMethod psiMethod,
                                             @NotNull String    methodName,
                                             @NotNull String    objectStreamClassName) {
		final String psiMethodName = psiMethod.getName();
		if (!methodName.equals(psiMethodName)) {
			return false;
		}

		final PsiParameterList parameterList = psiMethod.getParameterList();
		final PsiParameter[]   parameters    = parameterList.getParameters();

		if (parameters.length != 1) {
		    return false;
	    }

		final PsiType argType = parameters[0].getType();

		return (TypeUtils.typeEquals(objectStreamClassName, argType) &&
		        TypeUtils.typeEquals("void",                psiMethod.getReturnType()));
	}

	private static boolean isReadWriteResolve(@NotNull PsiMethod method, @NotNull String resolveMethodName) {
		final String methodName = method.getName();

		if (!resolveMethodName.equals(methodName)) {
	        return false;
	    }

		final PsiParameterList parameterList = method.getParameterList();
		final PsiParameter[]   parameters    = parameterList.getParameters();

		return (parameters.length == 0 &&
		        TypeUtils.isJavaLangObject(method.getReturnType()));
	}
}

File: src/main/kotlin/intellijplugin/util/SerializationUtils.kt
package intellijplugin.util

import com.intellij.psi.PsiClass
import com.intellij.psi.PsiMethod
import com.intellij.psi.util.InheritanceUtil
import org.jetbrains.kotlin.psi.KtClass
import org.jetbrains.kotlin.psi.KtClassOrObject
import org.jetbrains.kotlin.psi.KtSuperTypeListEntry
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

object SerializationUtils {
    private const val SERIALIZABLE_CLASS_NAME = "java.io.Serializable"
    private const val EXTERNALIZABLE_CLASS_NAME = "java.io.Externalizable"
    fun isSerializable(psiClass: PsiClass?): Boolean {
        if (psiClass == null) {
            return false
        }
        val serializable = ClassUtils.findPsiClass(
            psiClass.manager,
            SERIALIZABLE_CLASS_NAME
        )
        return InheritanceUtil.isInheritorOrSelf(psiClass, serializable, true)
    }

    fun isSerializable(psiClass: KtClass?): Boolean {
        return psiClass?.let { isImplementingSerializable(it) } ?: false
    }

    fun isImplementingSerializable(classOrObject: KtClassOrObject): Boolean {
        val list = classOrObject.superTypeListEntries
        val result = AtomicBoolean(false)
        list.forEach(Consumer { ktSuperTypeListEntry: KtSuperTypeListEntry ->
            if ("Serializable" == ktSuperTypeListEntry.text || "java.io.Serializable" == ktSuperTypeListEntry.text) {
                result.set(true)
            }
        })
        return result.get()
    }

    fun isExternalizable(psiClass: PsiClass): Boolean {
        val serializable = ClassUtils.findPsiClass(
            psiClass.manager,
            EXTERNALIZABLE_CLASS_NAME
        )
        return InheritanceUtil.isInheritorOrSelf(psiClass, serializable, true)
    }

    fun isDirectlySerializable(psiClass: PsiClass): Boolean {
        val implementsList = psiClass.implementsList
        if (implementsList != null) {
            for (aInterfaces in implementsList.referenceElements) {
                val implemented = aInterfaces.resolve() as PsiClass?
                if (implemented != null && SERIALIZABLE_CLASS_NAME == implemented.qualifiedName) {
                    return true
                }
            }
        }
        return false
    }

    fun hasReadObject(psiClass: PsiClass): Boolean {
        for (psiMethod in psiClass.methods) {
            if (isReadObject(psiMethod)) {
                return true
            }
        }
        return false
    }

    fun hasWriteObject(psiClass: PsiClass): Boolean {
        for (psiMethod in psiClass.methods) {
            if (isWriteObject(psiMethod)) {
                return true
            }
        }
        return false
    }

    fun isReadObject(psiMethod: PsiMethod): Boolean {
        return isReadWriteObject(
            psiMethod,
            "readObject",
            "java.io.ObjectInputStream"
        )
    }

    fun isWriteObject(psiMethod: PsiMethod): Boolean {
        return isReadWriteObject(
            psiMethod,
            "writeObject",
            "java.io.ObjectOutputStream"
        )
    }

    fun isReadResolve(method: PsiMethod): Boolean {
        return isReadWriteResolve(method, "readResolve")
    }

    fun isWriteReplace(method: PsiMethod): Boolean {
        return isReadWriteResolve(method, "writeReplace")
    }

    private fun isReadWriteObject(
        psiMethod: PsiMethod,
        methodName: String,
        objectStreamClassName: String
    ): Boolean {
        val psiMethodName = psiMethod.name
        if (methodName != psiMethodName) {
            return false
        }
        val parameterList = psiMethod.parameterList
        val parameters = parameterList.parameters
        if (parameters.size != 1) {
            return false
        }
        val argType = parameters[0].type
        return TypeUtils.typeEquals(objectStreamClassName, argType) &&
            TypeUtils.typeEquals("void", psiMethod.returnType)
    }

    private fun isReadWriteResolve(method: PsiMethod, resolveMethodName: String): Boolean {
        val methodName = method.name
        if (resolveMethodName != methodName) {
            return false
        }
        val parameterList = method.parameterList
        val parameters = parameterList.parameters
        return parameters.size == 0 &&
            TypeUtils.isJavaLangObject(method.returnType)
    }
}